util/dvsim/Scheduler.py

```diff
@@ -319,28 +317,40 @@ def _dispatch(self, hms):
         def __sum(d):
             return sum([len(d[k]) for k in d])
 
-        num_slots = min(Scheduler.max_parallel - __sum(self._running),
-                        __sum(self._queued))
-        if num_slots <= 0:
+        slots = min(Scheduler.max_parallel - __sum(self._running),
+                    __sum(self._queued))
+        if slots <= 0:
             return
 
+        # Compute how may slots to allocate to each target based on their
+        # weights.
+        sum_weight = 0
+        slots_filled = 0
+        total_weight = sum([
+            self._queued[t][0].weight for t in self._queued if self._queued[t]
+        ])
+
         for target in self._scheduled:
-            num_slots_per_target = min(
-                num_slots, Scheduler.max_parallel - len(self._running[target]))
-            if num_slots_per_target <= 0:
+            if not self._queued[target]:
+                continue
+
+            sum_weight += self._queued[target][0].weight
+            target_slots = round(
+                (slots * sum_weight) / total_weight) - slots_filled
```

Comment:
I'm probably just being a bit slow here, but I don't quite understand the weight computation. Would you mind adding a couple of comments to explain what we're calculating?

Reply:
Brill, thanks!